### PR TITLE
developer-docs of the earthly codebase

### PR DESCRIPTION
**Pull request by alexcb**
This is a rough starting-point for developers who want to learn more about the internals of the earthly codebase.

### CODE REVIEW

File: docs-internals/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+Earthly use BuildKit for executing commands, caching, and generating images. BuildKit contains similar contributor-focused documentation under
+the [buildkit/docs/dev](https://github.com/moby/buildkit/tree/master/docs/dev) section of their codebase.
+
+## Jargon
+
+| Name | Description |
+| :--- | :---------- |
+| **BuildKit** | BuildKit is a toolkit for converting source code to build artifacts in an efficient, expressive and repeatable manner. [^1] |
+| **LLB** | LLB is a BuildKit concept, which stands for "low-level build" definition[^2]; Earthly converts Earthfiles into LLB definitions, which is sent to BuildKit via the BuildKit client |
+| **LLB State** | LLB State, or simply State, is another BuildKit concept, which is used to produce low-level build devintions (LLB) from higher-level concepts like images, shell executions, mounts, etc [^2] |
```
Comment (vladaionescu) [inline docs-internals/README.md:22]:
```suggestion
| **LLB State** | LLB State, or simply State, is another BuildKit concept, which is used to produce low-level build definitions (LLB) from higher-level concepts like images, shell executions, mounts, etc [^2] |
```

File: docs-internals/README.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+| **pllb** | pllb is an Earthly thread-safe "parallel" wrapper around the LLB State functions |
+| **ast** | Abstract syntax tree; a custom Earthfile grammar is defined under the ast package, which [ANTLR](https://www.antlr.org/) uses to parse the initial Earthfile |
+| **buildcontext** | Borrowed from the `docker build --build-context` option, the buildcontext package ties the locations `COPY` reference to be relative to the corresponding Earthfile  |
+| **resolver** | The resolver takes an Earthly target (e.g. `./sub/dir+target`, or `github.com/...+target`), and constructs a llb state that points to the buildcontext   |
+| **builder** | The builder is the initial entrypoint, for the Earthly `build` cli command, it contains the `buildFunc` which is passed to the BuildKit client |
+| **interpreter** | The Earthly interpreter walks the ast, performing additional parsing and validation, and makes appropriate calls to the converter |
+| **converter** | The converter produces LLB, which is sent to BuildKit via the BuildKit client |
```
Comment (vladaionescu) [inline docs-internals/README.md:29]: Worth adding the "gateway" somewhere in here? The converter actually sends LLB via the gateway.

File: docs-internals/README.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+| **mts** | Multi-target states; which holds multiple LLB States, in the order they should be built |
+| **pullping** | PullPing is used to transfer images from BuildKit to host upon a `SAVE IMAGE`; the BuildKit client will call a onPull function, which will execute `docker pull 127.0.0.1:<port>/...` against a local registry (or proxy) |
```
Comment (vladaionescu) [inline docs-internals/README.md:31]:
```suggestion
| **pullping** | PullPing is used to transfer images from BuildKit to host upon a `SAVE IMAGE`; the BuildKit client will call a onPull function, which will execute `docker pull 127.0.0.1:<port>/...` against a local registry (or proxy). In other words, pullping is a signal sent by the Earthly exporter within Buildkit to the client (Earthly) to begin pulling images out of the embedded registry. |
```

Comment (vladaionescu) [reply docs-internals/README.md:31]: Maybe add definitions for the "Earthly exporter" within Buildkit, and the embedded registry (sometimes called local registry).

**Review by vladaionescu — COMMENTED** [review]
Looks good! Left a few small suggestions...

File: docs-internals/README.md (hunk 3 above)
Comment (alexcb) [reply docs-internals/README.md]: Thanks for the suggestion -- I tried rewording the two sentences to make them more connected.

**Review by vladaionescu — APPROVED** [review]

**Pull request merged**